utils/reusable_buffer.hh

```diff
@@ -8,123 +8,156 @@
-/// bytes_ostream potentially_fragmented_buffer;
-/// bytes_view view = rb.get_linearized_view(potentially_fragmented_buffer);
-/// // view is a view of a buffer that holds the same data as potentially_fragmented buffer
-/// ```
-///
-/// Example of writing to a bytes_ostream:
-///
-/// ```c++
-/// thread_local reusable_buffer rb;
-/// size_t maximum_size = compute_maximum_size();
-/// bytes_ostream destination = rb.make_buffer([&] (bytes_mutable_view view) {
-///     // view is a mutable view of some buffer which content will be in
-///     // the bytes_ostream returned by make_buffer.
-///     return actual_length_of_the_buffer;
-/// });
-/// ```
-class reusable_buffer { // extract to utils
-    // FIXME: We should start using std::byte for these things.
-    std::unique_ptr<int8_t[]> _buffer;
-    size_t _size = 0;
+template <typename Clock>
+class reusable_buffer_guard;
+
+/* Sometimes (e.g. for compression), we need a big contiguous buffer
+ * of a given size.
+ *
+ * Big buffers are a problem for the allocator, so if such a buffer
+ * is needed regularly, we want to reuse it. In the optimum, we
+ * only want to use one buffer per concurrent task (if all uses
+ * of the buffer are synchronous, one buffer per shard is enough),
+ * and only resize the buffer when it's too small for the current
+ * request.
+ *
+ * At the same time we might not want to keep the buffer around forever.
+ * A pathological request might cause the buffer to grow to some
+ * unreasonable size. After the source of the pathology (e.g. bad compression
+ * chunk size) is fixed, we want the buffer to free the capacity that will
+ * most likely not be used again.
+ *
+ * This class provides a buffer which balances both tasks by periodically
+ * shrinking its size to the largest recently seen size.
+ * The buffer always has a power-of-2 size, to minimize the number of
+ * allocations for small size changes.
+ *
+ * Even though the views returned by the buffer's methods have exactly the same
+ * requested size, the entire underlying capacity (of size size()) is safe to use.
+ * The returned views are always a prefix of the underlying capacity.
+ *
+ * Under a stable workload, the buffer's size will be stable.
+ * If a pathological request comes, the buffer will grow to accomodate it.
+ * 1-2 periods after the pathology ceases, the buffer will shrink back
+ * to its steady-state size.
+ *
+ * Be very careful with this buffer.
+ * Its content might disappear during a preemption point. This can be
+ * checked by comparing reallocs() before and after the preemption point.
+ * The content might be also easily invalidated by an unrelated future
+ * sharing the same buffer, if you are not careful.
+ *
+ * To make this buffer slightly harder to misuse, it's only accessible
+ * by reusable_buffer_guard.
+ * There can only exist one guard at a time.
+ * The guard can be used to obtain a view at most once.
+ */
+template <typename Clock>
+class reusable_buffer {
```

Comment:
To avoid over-templating here, let's make reusable_buffer a non-template and add a wrapper class that is a template on the clock, and only adds the timer functionality,

Reply:
Done. I hate splitting the class just to make the job easier for the compiler, though.